packages/main/src/Toolbar.ts

```diff
@@ -0,0 +1,497 @@
+			this.ITEMS_WIDTH_MAP.set(id, itemWidth);
+		});
+
+		this.ITEMS_WIDTH = totalWidth;
+		this.itemsWidthMeasured = true;
+	}
+
+	distributeItems(overflowSpace = 0) {
+		overflowSpace += this.OVERFLOW_BTN_SIZE;
+
+		this.itemsToBar = [];
+		this.itemsToOverflow = [];
+
+		// distribute items that always overflow
+		this.distributeItemsThatAlwaysOverflow();
+
+		// distribute the rest of the items, based on the available space
+		this.movableItems.reverse().forEach(item => {
+			if (overflowSpace > 0 && item.getAttribute("overflow-priority") !== ToolbarItemOverflowBehavior.NeverOverflow) {
+				this.itemsToOverflow.unshift(item);
+				overflowSpace -= this.getCachedItemWidth(item._id)!;
+			} else {
+				this.itemsToBar.unshift(item);
+			}
+		});
+
+		// If the last bar item is a spacer, force it to the overflow even if there is enough space for it
+		if (this.itemsToBar.length) {
+			const lastItemToBar = this.itemsToBar[this.itemsToBar.length - 1];
+			if (lastItemToBar.ignoreSpace) {
+				const itemBar = this.itemsToBar.pop();
+				if (itemBar) {
+					this.itemsToOverflow.unshift(itemBar);
+				}
+			}
+		}
+	}
+
+	displayAllItemsIntoBar() {
+		this.itemsToOverflow = [];
+
+		// distribute items that always overflow
+		this.distributeItemsThatAlwaysOverflow();
+
+		// distribute items that always overflow
+		this.distributeItemsThatNeverOverflow();
+
+		// distribute the rest of the items into the bar
+		this.itemsToBar = this.movableItems.map((item: ToolbarItem) => item);
+	}
+
+	distributeItemsThatAlwaysOverflow() {
+		this.alwaysOverflowItems.forEach((item: ToolbarItem) => {
+			this.itemsToOverflow.push(item);
+		});
+	}
+
+	distributeItemsThatNeverOverflow() {
+		this.neverOverflowItems.forEach((item: ToolbarItem) => {
+			this.itemsToBar.push(item);
+		});
+	}
+
+	get alwaysOverflowItems() {
+		return this._items.filter((item: ToolbarItem) => item.getAttribute("overflow-priority") === ToolbarItemOverflowBehavior.AlwaysOverflow);
+	}
+
+	get movableItems() {
+		return this._items.filter((item: ToolbarItem) => item.getAttribute("overflow-priority") !== ToolbarItemOverflowBehavior.AlwaysOverflow);
+	}
+
+	get neverOverflowItems() {
+		return this._items.filter((item: ToolbarItem) => item.getAttribute("overflow-priority") === ToolbarItemOverflowBehavior.NeverOverflow);
+	}
+
+	/**
+	 * Event Handlers
+	 */
+	onResize() {
+		if (!this.itemsWidthMeasured) {
+			return;
+		}
+
+		this.resizing = true;
+		this.closeOverflow();
+		this.processOverflowLayout();
+	}
+
+	onCustomItemClick(e: MouseEvent) {
+		const target = e.target as HTMLElement;
+		const item = target.closest<ToolbarItem>(".ui5-tb-item") || target.closest<ToolbarItem>(".ui5-tb-popover-item");
+
+		if (target === this.overflowButtonDOM) {
+			this.openOverflow();
+			return;
+		}
+
+		if (!item) {
+			return;
+		}
+
+		const refItemId = target.getAttribute("data-ui5-external-action-item-id");
+
+		if (refItemId) {
+			this.getItemByID(refItemId)!.fireEvent("click", {
```

Comment:
This code is in `Toolbar.ts` but is specific to toolbar buttons only, and is therefore hard-coded for them. Future items might not have a click event, or may have many. F.e. a dropdown will have a `change` event most likely, or both `change` and `click`.

There should be a generic `interact` or something of similar nature event on the toolbar itself, and all children should trigger this event on the toolbar with specific data, relevant to the toolbar item.